R/art.con.internal.R

```diff
@@ -0,0 +1,320 @@
+
+### Parses and validates contrast formula of form  "a:b:c"
+### Raises exception if formula does not validate.
+### returns list with:
+### interaction.variables: list of fixed variables (of type name) in interaction (e.g., list(a, b, c)).
+### interaction.term.labels: quoted interaction term label (e.g. "a:b:c")
+### concat.interaction.variable: concatenation (of type name) of all variables in interaction.variables (e.g., abc)
+#' @importFrom stats terms
+#' @importFrom pryr substitute_q
+#' @importFrom stringr str_replace
+parse.art.con.string.formula = function(f.orig){
+  
+  # check if f.orig is a single string (as opposed to a vector of multiple strings)
+  if(!is.character(f.orig) || length(f.orig) != 1){
+    stop("Contrast must be string term of form \"X1:X2\")")
+  }
+  
+  # don't allow spaces. must be "a:b", otherwise hard to parse
+  f.orig.has.space = grepl(' ', f.orig, fixed=TRUE)
+  if(f.orig.has.space){
+    stop("Contrast term cannot have a space in it.")
+  }
+  
+  # e.g. "a:b:c" -> list("a", "b", "c")
+  variables.str = strsplit(f.orig, ":")[[1]]
+  # e.g. list("a", "b", "c") -> list (a, b, c)
+  variables = lapply(variables.str, function(term) as.name(term))
```

Comment:
I always find it a little iffy to see manual parsing of code with string operations, especially in a language like R that can parse code for you. There are special cases of term names that are valid R code that won't be handled correctly by parsing the code manually (e.g., terms can theoretically include spaces or even `:` in their names as long as they are escaped properly).

To handle this, I would convert the interaction spec into an unevaluated expression and then use operations on the expression object to determine the information you are looking for. Something like this:

```r
interaction_string = "a:b:`var with spaces and : in it`"

# parse spec into an expression
interaction_expr = parse(text = interaction_string, keep.source = FALSE)[[1]]
interaction_expr
```
```
a:b:`var with spaces and : in it`
```
```r
# recursively traverse the interection spec to extract variables
# and to validate that it only uses `:` as its only operator
get_variables = function(spec) {
  if (is.call(spec) && spec[[1]] == ":") {
    # recursive case: get variables from children
    c(get_variables(spec[[2]]), get_variables(spec[[3]]))
  } else if (is.name(spec)) {
    # base case: this is a variable
    spec
  } else {
    stop("insert some meaningful error about how interaction specs can only contain variables and `:`s")
  }
}

variables = get_variables(interaction_expr)
variables
```
```
[[1]]
a

[[2]]
b

[[3]]
`var with spaces and : in it`
```